### ISSUE TITLE
`manual_flatten` false positives with reference slices

### ISSUE THREAD

**Issue by awused** [rust-lang/rust-clippy#7514]
Lint name: manual_flatten

I tried this code:

```rust
struct Test {
    a: Option<usize>,
    b: Option<usize>,
}

fn main() {
    let t = Test {
        a: Some(1),
        b: Some(2),
    };

    for x in [&t.a, &t.b] {
        if let Some(y) = x {
            println!("{}", y);     
        }
    }
} 
```

This also applies to mutable references.

Clippy suggests this, which is both incomplete as a refactor and does not compile. Try as I might I could not get into_iter().flatten() to compile.

```rust
for x in [&t.a, &t.b].into_iter().flatten() {
    if let Some(y) = x {
        println!("{}", y);
    }
}
```

### Meta

I saw https://github.com/rust-lang/rust-clippy/issues/6893 which appears to be the same issue, but it was claimed to be fixed there and that fix should have landed in 1.53. I'm on 1.54 now and still seeing it with both mutable and immutable reference slices.

- `cargo clippy -V`: clippy 0.1.54 (a178d03 2021-07-26)
- `rustc -Vv`:
```
rustc 1.54.0 (a178d0322 2021-07-26)
binary: rustc
commit-hash: a178d0322ce20e33eac124758e837cbd80a6f633
commit-date: 2021-07-26
host: x86_64-unknown-linux-gnu
release: 1.54.0
LLVM version: 12.0.1
```

**Comment by dswij** [comment]
In addition to the original issue, `manual_flatten` also seems to give false suggestion on

```rust
fn main() {
    for n in [&&Some(1)] {
        if let Some(x) = n {
            println!("{}", x);
        }
    }
}
```

Running clippy suggests this which does not compile:

```rust
for n in [&&Some(1)].into_iter().copied().flatten() {
    println!("{}", n);
}
```

Problem lies on not checking nested reference on the lint [here](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/loops/manual_flatten.rs#L58-L64).  And the original issue suggested that the lint do not check for type `ty::Adt(..)`.

Using `flatten()` on nested references is a pain, since we would call `copied()` a lot of times. Would this lint makes more sense if it exempts reference altogether rather than suggesting `into_iter().copied().flatten()`?

[Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=fn%20main()%20%7B%0A%20%20%20%20%2F%2F%20manual_flatten%20flags%20this%0A%20%20%20%20for%20n%20in%20%5B%26%26Some(1)%5D%20%7B%0A%20%20%20%20%20%20%20%20if%20let%20Some(x)%20%3D%20n%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20println!(%22%7B%7D%22%2C%20x)%3B%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%20%20%0A%20%20%20%20%2F%2F%20and%20suggest%20this%20instead%2C%20which%20does%20not%20compile!%3A%0A%20%20%20%20%0A%20%20%20%20%2F%2F%20for%20n%20in%20%5B%26%26Some(1)%5D.into_iter().copied().flatten()%20%7B%0A%20%20%20%20%2F%2F%20%20%20%20%20println!(%22%7B%7D%22%2C%20n)%3B%0A%20%20%20%20%2F%2F%20%7D%0A%20%20%20%20%0A%20%20%20%20%2F%2F%20---%0A%20%20%20%20%0A%20%20%20%20%2F%2F%20This%20compiles%20if%20we%20want%20to%20use%20flatten()%2C%20but%20at%20what%20cost%3F%3A%0A%20%20%20%20%0A%20%20%20%20%2F%2F%20for%20n%20in%20%5B%26%26Some(1)%5D.into_iter().copied().copied().flatten()%20%7B%0A%20%20%20%20%2F%2F%20%20%20%20%20println!(%22%7B%7D%22%2C%20n)%3B%0A%20%20%20%20%2F%2F%20%7D%0A%7D)

**Comment by AppBlitz** [comment]
hello i like take issue, still available

**Comment by Twil3akine** [comment]
I tried to reproduce this on the current Clippy.

## Original example

```rust
struct Test {
    a: Option<usize>,
    b: Option<usize>,
}

fn main() {
    let t = Test {
        a: Some(1),
        b: Some(2),
    };

    for x in [&t.a, &t.b] {
        if let Some(y) = x {
            println!("{}", y);
        }
    }
}
```

`cargo clippy` still emits `manual_flatten`:

```text
warning: unnecessary `if let` since only the `Some` variant of the iterator element is used
  --> src/main.rs:12:5
   |
12 | /     for x in [&t.a, &t.b] {
13 | |         if let Some(y) = x {
14 | |             println!("{}", y);
15 | |         }
16 | |     }
   | |_____^
   |
help: try `.flatten()` and remove the `if let` statement in the for loop
  --> src/main.rs:13:9
   |
13 | /         if let Some(y) = x {
14 | |             println!("{}", y);
15 | |         }
   | |_________^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_flatten
   = note: `#[warn(clippy::manual_flatten)]` on by default
help: try
   |
12 ~     for y in [&t.a, &t.b].into_iter().flatten() {
13 +         println!("{}", y);
14 +     }
   |
```

Then I ran:

```text
cargo clippy --fix --allow-dirty
cargo check
```

Result:

```text
Checking clippy-msrv-test v0.1.0 (/path/to/clippy-msrv-test)
   Fixed src/main.rs (1 fix)
Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
Checking clippy-msrv-test v0.1.0 (/path/to/clippy-msrv-test)
Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.03s
```

So the suggestion is applied successfully, and the code compiles afterwards.

## Nested reference example from the comments

```rust
fn main() {
    for n in [&&Some(1)] {
        if let Some(x) = n {
            println!("{}", x);
        }
    }
}
```

`cargo clippy` emits `manual_flatten`:

```text
warning: unnecessary `if let` since only the `Some` variant of the iterator element is used
 --> src/main.rs:2:5
  |
2 | /     for n in [&&Some(1)] {
3 | |         if let Some(x) = n {
4 | |             println!("{}", x);
5 | |         }
6 | |     }
  | |_____^
  |
help: try `.flatten()` and remove the `if let` statement in the for loop
 --> src/main.rs:3:9
  |
3 | /         if let Some(x) = n {
4 | |             println!("{}", x);
5 | |         }
  | |_________^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_flatten
  = note: `#[warn(clippy::manual_flatten)]` on by default
help: try
  |
2 ~     for x in [&&Some(1)].into_iter().copied().flatten() {
3 +         println!("{}", x);
4 +     }
  |
```

Then I ran:

```text
cargo clippy --fix --allow-dirty
cargo check
```

Result:

```text
Checking clippy-msrv-test v0.1.0 (/path/to/clippy-msrv-test)
   Fixed src/main.rs (1 fix)
Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
Checking clippy-msrv-test v0.1.0 (/path/to/clippy-msrv-test)
Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.03s
```

So this suggestion is also applied successfully, and the code compiles afterwards.

I could not reproduce the broken suggestion described in this issue on current Clippy with the examples above. The lint is still emitted, but the suggested fix appears to be complete and compiles successfully now.

Could someone confirm whether this is still reproducible, or whether this issue can be closed?

**Comment by awused** [comment]
It seems to be fixed, the bug was about a bad suggestion (and possibly an invalid lint for old versions of rust but I can't remember the details) so now that the bad suggestion isn't bad the bug can be closed.

**Comment by samueltardieu** [comment]
The difference you see is that Rust 1.54, the original compiler you tried it on, uses Rust edition 2018. The behavior of `slice::into_iter()` changed in Rust edition 2021, which is why the fix proposed by Clippy now seems correct.

However, as exhibited by @dswij in <https://github.com/rust-lang/rust-clippy/issues/7514#issuecomment-889822717>, the issue (not checking that the `Item` of the outer iterator implements `IntoIterator`) is still present. Reopening.

**Comment by samueltardieu** [comment]
> Could someone confirm whether this is still reproducible, or whether this issue can be closed?

It is still reproducible, but you have to force using the 2018 edition of Rust to see it, or use the other example in <https://github.com/rust-lang/rust-clippy/issues/7514#issuecomment-889822717> which shows the bug on any edition.

**Comment by Twil3akine** [comment]
Thanks for the clarification. I missed the edition-dependent behavior of `slice::into_iter()` and only tested with the current default setup.

I’ll re-check this with edition 2018, and also test the example from the linked comment that is supposed to reproduce independently of the edition.

**Comment by Twil3akine** [comment]
Thanks for the clarification. I re-tested this with explicit editions.

I tested both:

1. the original example from the issue
2. the nested reference example from the linked comment

using:

```bash
cargo clippy
cargo clippy --fix --allow-dirty
cargo check
```

## Original example

```rust
struct Test {
    a: Option<usize>,
    b: Option<usize>,
}

fn main() {
    let t = Test {
        a: Some(1),
        b: Some(2),
    };

    for x in [&t.a, &t.b] {
        if let Some(y) = x {
            println!("{}", y);
        }
    }
}
```

With `edition = "2015"` and `edition = "2018"`, this is still reproducible.

Clippy suggests:

```rust
for y in [&t.a, &t.b].into_iter().flatten() {
    println!("{}", y);
}
```

but `cargo clippy --fix --allow-dirty` fails because the fixed code does not compile:

```text
error[E0277]: `&&std::option::Option<usize>` is not an iterator
```

With `edition = "2021"`, the same suggestion is applied successfully and `cargo check` passes.

## Nested reference example

```rust
fn main() {
    for n in [&&Some(1)] {
        if let Some(x) = n {
            println!("{}", x);
        }
    }
}
```

With `edition = "2015"` and `edition = "2018"`, this is also reproducible.

Clippy suggests:

```rust
for x in [&&Some(1)].into_iter().copied().flatten() {
    println!("{}", x);
}
```

but `cargo clippy --fix --allow-dirty` fails because the fixed code does not compile:

```text
error[E0277]: `&&std::option::Option<{integer}>` is not an iterator
```

With `edition = "2021"`, the suggestion is applied successfully and `cargo check` passes.

So the issue is still reproducible on current Clippy, but from my tests it appears to affect editions before 2021. My previous comment was incomplete because I only tested the current default setup and did not account for the pre-2021 behavior of array `into_iter()`.

**Comment by samueltardieu** [comment]
> Thanks for the clarification. I re-tested this with explicit editions. […]

Has this response been generated by a LLM? If this is the case, please abstain doing this, we want to interact with human contributors here, not with machines. If you use a LLM on your side to understand an issue, this is fine, but please don't copy'n-paste a (lengthy) LLM answer here.

**Comment by Twil3akine** [comment]
> Has this response been generated by a LLM?

Yes, That is true...
I apologize for using LLM to reply. I'm not very good at English, so I requested LLM for generation reply. 

I propose to write replies myself from now.
If my reply is hard to read, please tell me...orz

this reply is written by me (but using [`Nani translation`](https://nani.now/en)).
